src/charm.py

```diff
@@ -1167,6 +1205,33 @@ def generate_auth_env_vars(self, conf: Dict[str, Dict[str, Any]]) -> Dict[str, s
             env_vars[auth_var_prefix + str(var).upper()] = str(conf[auth_mode][var])
         return env_vars
 
+    @property
+    def _client_config(self) -> ClientConfig:
+        return ClientConfig(
+            urljoin(self.external_url, "/login/generic_oauth"),
+            OAUTH_SCOPES,
+            OAUTH_GRANT_TYPES,
+        )
+
+    def _set_client_config(self, relation_id: int) -> None:
+        self.oauth.update_client_config(client_config=self._client_config, relation_id=relation_id)
+
+    def _on_oauth_info_changed(self, event: OAuthInfoChangedEvent) -> None:
+        """Event handler for the oauth_info_changed event."""
+        if not self.unit.is_leader():
+            return
+
+        relation_id = self.model.relations[OAUTH][0].id
+        self._set_client_config(relation_id)
+        logger.info(f"Received oauth provider info: {self.oauth.get_provider_info(relation_id)}")
+
+        if self.oauth.is_client_created(relation_id):
+            self.restart_grafana()
+        else:
+            logger.info("No oauth client info available, deferring the event")
+            event.defer()
+            return
```

Comment:
I would re-write this `if...else` so we can remove the `else` and it is clearer to read

```suggestion
        if not self.oauth.is_client_created(relation_id):
            logger.info("No oauth client info available, deferring the event")
            event.defer()
            return

        self.restart_grafana()
```